### PR TITLE
Integrate TensorRT into GPU Python package

### PR DESCRIPTION
**Pull request by chilo-ms**
Add TRT into Python GPU package alongside with CUDA. We also disable TRT EP and only run CUDA EP in ONNX backend test to retain previous behavior.

### CODE REVIEW

File: onnxruntime/python/backend/backend.py [hunk 1]
```diff
@@ -107,7 +107,19 @@ def prepare(cls, model, device=None, **kwargs):
...
+
+            providers = get_available_providers()
+
+            if os.getenv('EXCLUDE_PROVIDERS'):
+                exclude_providers = os.getenv('EXCLUDE_PROVIDERS').split(',')
+            else:
+                exclude_providers = []
+
+            # exclude TRT EP temporarily and only test CUDA EP to retain previous behavior
+            if 'TensorrtExecutionProvider' in providers and 'TensorrtExecutionProvider' in exclude_providers:
```
Comment (jywu-msft) [inline onnxruntime/python/backend/backend.py:119]: remove this entire block for checking TRT EP. only exclude via EXCLUDE_PROVIDERS

File: onnxruntime/python/backend/backend.py [hunk 2]
```diff
@@ -107,7 +107,19 @@ def prepare(cls, model, device=None, **kwargs):
             for k, v in kwargs.items():
                 if hasattr(options, k):
                     setattr(options, k, v)
-            inf = InferenceSession(model, sess_options=options, providers=get_available_providers())
+
+            providers = get_available_providers()
+
+            if os.getenv('EXCLUDE_PROVIDERS'):
```
Comment (jywu-msft) [inline onnxruntime/python/backend/backend.py:113]: i think we need a better name.

File: onnxruntime/python/backend/backend.py (hunk 1 above)
Comment (jywu-msft) [reply onnxruntime/python/backend/backend.py:119]: the logic should be to remove excluded providers from the list returned by get_available_providers()
then you don't need to hardcode or specifically mention any providers by name.

File: setup.py [hunk 3]
```diff
@@ -135,6 +133,17 @@ def _rewrite_ld_preload_cuda(self, to_preload):
...
+        def _rewrite_ld_preload_tensorrt(self, to_preload):
+            with open('onnxruntime/capi/_ld_preload.py', 'a') as f:
+                if len(to_preload) > 0:
+                    f.write('from ctypes import CDLL, RTLD_GLOBAL\n')
+                    f.write('try:\n')
+                    for library in to_preload:
+                        f.write('    _{} = CDLL("{}", mode=RTLD_GLOBAL)\n'.format(library.split('.')[0], library))
+                    f.write('except OSError:\n')
+                    f.write('    import os\n')
+                    f.write('    os.environ["ORT_TENSORRT_UNAVAILABLE"] = "1"\n')
```
Comment (jywu-msft) [inline setup.py:145]: pybind_state.cc needs to be modified to check if ORT_TENSORRT_UNAVAILABLE if set to 1? 
same as what's being done for ORT_CUDA_UNAVAILABLE

Comment (chilo-ms) [reply setup.py:145]: yes, that's right. Will add in pybind_state.cc

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 4]
```diff
@@ -352,151 +352,157 @@ std::unique_ptr<IExecutionProvider> CreateExecutionProviderInstance(
...
+          return onnxruntime::CreateExecutionProviderFactory_Tensorrt(&params)->CreateProvider();
         } else {
-          ORT_THROW("Invalid TensorRT EP option: ", option.first);
+          return onnxruntime::CreateExecutionProviderFactory_Tensorrt(cuda_device_id)->CreateProvider();
         }
-      }
-      return onnxruntime::CreateExecutionProviderFactory_Tensorrt(&params)->CreateProvider();
     } else {
-      return onnxruntime::CreateExecutionProviderFactory_Tensorrt(cuda_device_id)->CreateProvider();
+      ORT_THROW("Please install the correct version TensorRT as well as CUDA and cuDNN as mentioned in the TensorRT requirements page (https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#requirements), make sure they're in the PATH, and that your GPU is supported.");
```
Comment (jywu-msft) [inline onnxruntime/python/onnxruntime_pybind_state.cc:505]: We do not want to unconditionally throw here.
this else condition can be hit on a system without cuda/tensorrt installed.
I think we should follow the cuda example and check for whether CUDA_PATH is defined as a hint that this is a system with nvidia gpu/cuda installed.

Comment (jywu-msft) [reply onnxruntime/python/onnxruntime_pybind_state.cc:505]: can you please test the package on a system with just CUDA , TRT + CUDA and one without CUDA to confirm all those scenarios work properly. thanks!

Comment (chilo-ms) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Rewrote the condition as you suggested to throw exception.
Will test the package once it's been built by CI Pipelines.

File: onnxruntime/python/backend/backend.py [hunk 5]
```diff
@@ -107,7 +107,16 @@ def prepare(cls, model, device=None, **kwargs):
             for k, v in kwargs.items():
                 if hasattr(options, k):
                     setattr(options, k, v)
-            inf = InferenceSession(model, sess_options=options, providers=get_available_providers())
+
+            providers = get_available_providers()
+
+            if os.getenv('ORT_ONNX_BACKEND_EXCLUDE_PROVIDERS'):
```
Comment (jywu-msft) [inline onnxruntime/python/backend/backend.py:113]: would be more compact to write it like this:

excluded_providers = os.getenv('ORT_ONNX_BACKEND_EXCLUDE_PROVIDERS', default="").split(',')
providers = [x for x in providers if (x not in excluded_providers)]

File: onnxruntime/python/backend/backend.py [hunk 6]
```diff
@@ -107,7 +107,16 @@ def prepare(cls, model, device=None, **kwargs):
             for k, v in kwargs.items():
                 if hasattr(options, k):
                     setattr(options, k, v)
-            inf = InferenceSession(model, sess_options=options, providers=get_available_providers())
+
+            providers = get_available_providers()
```
Comment (jywu-msft) [inline onnxruntime/python/backend/backend.py:111]: would be more compact to rewrite lines 111-117 as

excluded_providers = os.getenv('ORT_ONNX_BACKEND_EXCLUDE_PROVIDERS', default="").split(',')
providers = [x for x in get_available_providers() if (x not in excluded_providers)]

Comment (chilo-ms) [reply onnxruntime/python/backend/backend.py:111]: okay, same logic but more compact

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**